File: disasm/avg.py
"""Atari Analog Vector Generator (AVG) decoder/encoder for Tempest.

Adapted from the Space Duel disassembly's avg.py. Tempest-specific semantics
were checked against:

  * tempest-main/VGMC.MAC (Ed Logg's macros: VCTR, SVEC short form, STAT, SCAL,
    CNTR, HALT, JSRL, RTSL, JMPL) and ALVROM.MAC (CSTAT = .WORD 68C0+colour)
  * the AAE/MAME AVG core used by tempest.cpp (mame_late_avgdvg.cpp,
    avg_tempest: tempest_strobe2 / tempest_strobe3)

Opcode = word >> 13 (the first word, little-endian in the ROM):

  0 VCTR  2 words: w1 = DY & $1FFF, w2 = Z<<13 | DX & $1FFF  (13-bit two's
          complement deltas, NOT sign-magnitude as on the DVG)
  1 HALT  $2000
  2 SVEC  short vector $4000 | Z<<5 | (DX/2)&$1F | ((DY/2)&$1F)<<8
  3 STAT  $6xxx when bit 12 clear.  TEMPEST: if bit 11 ($0800) is set the low
          nibble loads the COLOUR latch (index into colour RAM $0800-$080F);
          otherwise (w>>4)&$F loads the INTENSITY latch.  ALVROM's CSTAT c
          writes $68C0+c; ALDIS2 calls VGSTAT with MZCOLO=8 (colour) or
          MZBRIT=0 (intensity).  Tempest has no sparkle / X-flip bits: those
          ($0800 sparkle, $0400 flip) belong to Major Havoc's AVG
          (mhavoc_strobe2); on Tempest the $0400/$0200/$0100 bits that VGMC's
          STAT macro can set (HI/LOW, IN/OUT window) are ignored.
      SCAL  $7000 | b<<8 | linear  (bit 12 set): binary scale b (0..7),
          linear scale l (factor (255-l)/256 ... $00 = full, $80 = 1/2)
  4 CNTR  $8040 - centre the beam
  5 JSRL  $A000 | target/2  (VG address space: CPU $2000 + 2*(w & $1FFF))
  6 RTSL  $C000  (ALVROM hides checksum bytes in RTSL low bytes: CHKSM0/1)
  7 JMPL  $E000 | target/2

Vector intensity (tempest_strobe3): z = top 3 bits of the vector's X word
(SVEC: bits 7-5 of the low byte). z == 1 means "use the STAT intensity",
otherwise the brightness is z*2 (0 = beam off). Colour is always the last
colour STAT.
"""
import os

VG_BASE = 0x2000
HERE = os.path.dirname(os.path.abspath(__file__))
ROM64K = os.path.join(HERE, "_survey", "roms_extracted", "tempest3_cpu64k.bin")
VROM_LO, VROM_HI = 0x3000, 0x3FFF


def load_image(path=ROM64K):
    """64K CPU image as bytes (index = CPU address)."""
    data = open(path, "rb").read()
    assert len(data) == 0x10000, path
    return data


def s13(v):
    """13-bit two's complement delta (VGMC: .WORD DY&^H1FFF)."""
    v &= 0x1FFF
    return v - 0x2000 if v & 0x1000 else v


def s5(v):
    v &= 0x1F
    return v - 0x20 if v & 0x10 else v


def word(mem, a):
    return mem[a] | (mem[a + 1] << 8)


def vg_target(w):
    return VG_BASE + ((w & 0x1FFF) * 2)


def decode_raw(mem, a):
    """-> dict(op, len, fields...) without text formatting."""
    w = word(mem, a)
    top = w >> 13
    if top == 0:
        w2 = word(mem, a + 2)
        return dict(op="VCTR", len=4, dx=s13(w2), dy=s13(w), z=(w2 >> 13) & 7,
                    words=[w, w2])
    if top == 1:
        return dict(op="HALT", len=2, words=[w])
    if top == 2:
        return dict(op="SVEC", len=2, dx=s5(w) * 2, dy=s5(w >> 8) * 2,
                    z=(w >> 5) & 7, words=[w])
    if top == 3:
        if w & 0x1000:
            return dict(op="SCAL", len=2, b=(w >> 8) & 0xF, l=w & 0xFF, words=[w])
        if (w & 0x0F00) == 0x0800 and (w & 0xF0) == 0xC0:
            return dict(op="CSTAT", len=2, color=w & 0xF, words=[w])
        return dict(op="STAT", len=2, words=[w],
                    color=(w & 0xF) if w & 0x0800 else None,
                    intensity=None if w & 0x0800 else (w >> 4) & 0xF)
    if top == 4:
        return dict(op="CNTR", len=2, words=[w])
    if top == 5:
        return dict(op="JSRL", len=2, target=vg_target(w), words=[w])
    if top == 6:
        return dict(op="RTSL", len=2, words=[w])
    return dict(op="JMPL", len=2, target=vg_target(w), words=[w])


def decode(mem, a):
    """-> (mnemonic, operand_text, length) in Space Duel listing style."""
    d = decode_raw(mem, a)
    op = d["op"]
    if op in ("VCTR", "SVEC"):
        return op, "%d, %d, %d" % (d["dx"], d["dy"], d["z"]), d["len"]
    if op == "SCAL":
        return op, "%d, $%02X" % (d["b"], d["l"]), 2
    if op == "CSTAT":
        return op, "%d" % d["color"], 2
    if op == "STAT":
        return op, "$%04X" % d["words"][0], 2
    if op in ("JSRL", "JMPL"):
        return op, "$%04X" % d["target"], 2
    if op == "CNTR" and d["words"][0] != 0x8040:
        return op, "$%04X" % d["words"][0], 2
    if op == "RTSL" and d["words"][0] != 0xC000:
        return op, "$%04X" % d["words"][0], 2
    return op, "", d["len"]


def _ints(s):
    out = []
    for x in s.split(","):
        x = x.strip()
        out.append(int(x[1:], 16) if x.startswith("$") else int(x))
    return out


def encode(mn, oper):
    """Inverse of decode(): (mnemonic, operand text) -> list of 16-bit words,
    or None when the text cannot be encoded. JSRL/JMPL operands must be $hex
    (resolve labels before calling)."""
    oper = (oper or "").strip()
    if mn == "HALT":
        return [0x2000]
    if mn == "RTSL":
        return [_ints(oper)[0]] if oper else [0xC000]
    if mn == "CNTR":
        return [_ints(oper)[0]] if oper else [0x8040]
    if mn == "VCTR":
        dx, dy, z = _ints(oper)
        return [dy & 0x1FFF, ((z & 7) << 13) | (dx & 0x1FFF)]
    if mn == "SVEC":
        dx, dy, z = _ints(oper)
        if dx & 1 or dy & 1:
            return None
        return [0x4000 | ((z & 7) << 5) | ((dx >> 1) & 0x1F) | (((dy >> 1) & 0x1F) << 8)]
    if mn == "SCAL":
        b, l = _ints(oper)
        return [0x7000 | ((b & 0xF) << 8) | (l & 0xFF)]
    if mn == "CSTAT":
        return [0x68C0 | (_ints(oper)[0] & 0xF)]
    if mn == "STAT":
        return [_ints(oper)[0]]
    if mn in ("JSRL", "JMPL"):
        t = _ints(oper)[0]
        if t & 1 or not (VG_BASE <= t < VG_BASE + 0x4000):
            return None
        return [(0xA000 if mn == "JSRL" else 0xE000) | ((t - VG_BASE) >> 1)]
    return None


def scal_factor(b, lin):
    """aae_avg: timer shifted b times, linear scale complemented:
    factor = ((~lin)&0xFF)/256 / 2^b  (SCAL 0,$00 ~ full size)."""
    return ((~lin) & 0xFF) / 256.0 / (1 << (b & 7))


def targets(mem, lo=VROM_LO, hi=VROM_HI):
    out, a = set(), lo
    while a < hi:
        d = decode_raw(mem, a)
        if d["op"] in ("JSRL", "JMPL") and lo <= d["target"] <= hi:
            out.add(d["target"])
        a += d["len"]
    return out


if __name__ == "__main__":
    mem = load_image()
    counts, a = {}, VROM_LO
    while a < VROM_HI:
        d = decode_raw(mem, a)
        counts[d["op"]] = counts.get(d["op"], 0) + 1
        a += d["len"]
    print("AVG opcode census over $3000-$3FFF (linear sweep):")
    for k, v in sorted(counts.items(), key=lambda x: -x[1]):
        print("   %-6s %5d" % (k, v))
    print("JSRL/JMPL targets inside vector ROM: %d" % len(targets(mem)))

File: c_src/harvest/avg.c
/* avg.c - Space Duel C port: the Analog Vector Generator state machine.
 *
 * Transcribed from the generic USE_AVG path of avg_video_update() in
 * AAE_publish_new_vector_test\aae\aae\vidhrdwr\aae_avg.cpp, the same
 * reference ../disasm/avg.py was verified against (two's-complement deltas,
 * SCAL complement rule, z==1 = STAT intensity). Memory map per DESIGN.md
 * "JMPL address fold (RESOLVED)": flat word addressing from CPU $2000,
 * vector RAM words $000-$3FF, ROM words $400-$FFF, no mirror.
 *
 * Platform-free: only stdint/stddef plus the two project headers.
 */
#include <stdint.h>
#include <stddef.h>
#include "sd_state.h"
#include "sd_vecrom.h"
#include "avg.h"

/* Opcode = word >> 13 (SCAL is STAT with bit 12 set, split at dispatch). */
#define OP_VCTR 0
#define OP_HALT 1
#define OP_SVEC 2
#define OP_STAT 3
#define OP_CNTR 4
#define OP_JSRL 5
#define OP_RTSL 6
#define OP_JMPL 7

#define AVG_MAXSTACK   8        /* aae_avg.cpp MAXSTACK (hardware has 4)   */
#define AVG_MAXFETCH   0x8000   /* runaway guard: 32K word fetches is >4x  *
                                 * the whole addressable AVG space         */

/* Result registers of the most recent run (hardware-status analogue). */
static avg_stop avg_stop_reason;
static int      avg_flags;
static double   avg_time_units;
static uint32_t avg_cycles;

/* ---- cycle-true draw-time model (mame_late_avgdvg.cpp) --------------------
 *
 * The AVG is a state machine clocked from MASTER_CLOCK = 12,096,000 Hz
 * (mame_late_avgdvg.cpp line 45); every state-PROM tick costs 8 master
 * cycles (run_state_machine, line 1271 "cycles += 8"), i.e. the PROM is
 * stepped at 1.512 MHz.  Vector/center durations are charged on top by
 * avg_common_strobe3 (lines 934-965): a VCTR runs 0x8000 - timer master
 * cycles, an SVEC 0x100 - (timer & 0xFF), a CNTR 0x8000 - timer, where
 * `timer` is the shift register filled by the normalizer (avg_strobe0,
 * lines 660-698) and the binary-scale shifter (avg_strobe1, lines 740-756).
 *
 * The per-opcode PROM tick counts below were extracted by tracing the real
 * state PROM (136002-125.n4, ROM_REGION REGION_PROMS of the whole
 * bwidow/spacduel family - drivers/bwidow.cpp line 568) through
 * avg_state_addr()/run_state_machine() with the avg_default handler table
 * (mame_late_avgdvg.cpp lines 1407-1423 - the plain AVG variant spacduel
 * uses via avg_start(), line 1532; no clipping hardware):
 *
 *   op  ticks  states traversed          extra master cycles
 *   VCTR  8    9,8,B,A,C,D,F,0           0x8000 - timer
 *   HALT  2    9,8 (halt visible at F)   -
 *   SVEC  6    9,B,C,D,F,0               0x100 - (timer & 0xFF)
 *   STAT  7    9,8,E,3,2,1,0             -
 *   SCAL  7    9,8,E,3,2,1,0             -
 *   CNTR  5    9,8,C,F,0                 0x8000 - timer
 *   JSRL  5    9,8,C,D,E                 -
 *   RTSL  4    9,8,D,E                   -
 *   JMPL  3    9,8,E                     -
 *
 * plus one idle tick after VGGO before the first latch state.  This per-op
 * accounting reproduces the full PROM-driven state machine cycle-for-cycle
 * on every captured attract frame (scratchpad avg_cycle_sim.py, 2026-08-26).
 */
#define AVG_MASTER_HZ 12096000.0

/* Normalizer shift count (avg_strobe0): both 13-bit delta registers are
 * shifted up until either has its bit 12 differing from bit 11 (so the
 * DAC's bits 3-12 carry real information); dvx==dvy==0 cuts off at 16. */
static int avg_norm_shifts(uint16_t dvy, uint16_t dvx)
{
    int i = 0;

    while (((dvy ^ (uint16_t)(dvy << 1)) & 0x1000) == 0 &&
           ((dvx ^ (uint16_t)(dvx << 1)) & 0x1000) == 0 && i < 16) {
        i++;
        dvy = (uint16_t)((dvy & 0x1000) | ((dvy << 1) & 0x1FFF));
        dvx = (uint16_t)((dvx & 0x1000) | ((dvx << 1) & 0x1FFF));
    }
    return i;
}

/* The timer shift register after nnorm normalizer shifts (avg_strobe0) and
 * nbin binary-scale shifts (avg_strobe1). op1 = the SVEC path (OP1 set):
 * each shift also ORs 0x80 in and the register is masked to its low byte. */
static uint16_t avg_timer_reg(int nnorm, int nbin, int op1)
{
    uint16_t t = 0;
    int      i;

    for (i = 0; i < nnorm; i++) {
        t >>= 1;
        t |= (uint16_t)(op1 ? 0x4080 : 0x4000);
    }
    if (op1)
        t &= 0xFF;
    for (i = 0; i < nbin; i++) {
        t >>= 1;
        t |= (uint16_t)(op1 ? 0x4080 : 0x4000);
    }
    if (op1)
        t &= 0xFF;
    return t;
}

/* 13-bit two's complement (VGMC.MAC .WORD DY&^H1FFF; aae twos_comp_val).
 * NOT sign-magnitude - that is the older DVG's convention. */
static int s13(uint16_t v)
{
    v &= 0x1FFF;
    return (v & 0x1000) ? (int)v - 0x2000 : (int)v;
}

/* 5-bit two's complement half-delta of the short vector. */
static int s5(uint16_t v)
{
    v &= 0x1F;
    return (v & 0x10) ? (int)v - 0x20 : (int)v;
}

/* Fetch the little-endian word at AVG word address wa (flat map).
 * Returns 0 and stops the caller when wa is past the carried ROMs. */
static int avg_fetch(uint16_t wa, uint16_t *out)
{
    uint32_t off = ((uint32_t)wa & 0x1FFFu) << 1;   /* byte offset from $2000 */

    if (off < 0x0800u) {                            /* CPU $2000-$27FF: RAM   */
        *out = (uint16_t)(g.vram[off] | ((uint16_t)g.vram[off + 1] << 8));
        return 1;
    }
    if (off < 0x2000u) {                            /* CPU $2800-$3FFF: ROM   */
        if (off < 0x1000u)
            avg_flags |= AVG_FLAG_SHIPROM;          /* $2800-$2FFF: never     *
                                                     * legitimate AVG data    */
        *out = (uint16_t)(sd_vecrom[off - 0x0800u]
                          | ((uint16_t)sd_vecrom[off - 0x0800u + 1] << 8));
        return 1;
    }
    *out = 0;                                       /* program ROM: not here  */
    return 0;
}

void avg_run_from(uint16_t word_addr,
                  void (*seg)(float x0, float y0, float x1, float y1,
                              int color, int lum),
                  int *out_words_walked)
{
    uint16_t stack[AVG_MAXSTACK];
    int      sp = 0;
    uint16_t pc = (uint16_t)(word_addr & 0x1FFF);
    int      fetched = 0;

    double x = 0.0, y = 0.0;        /* beam at center, like the reference  */
    double scale = 0.0;             /* aae starts scale 0: pre-SCAL vectors*
                                     * have zero length                    */
    int statz = 0;                  /* STAT intensity register             */
    int color = 0;
    int bin_scale = 0;              /* SCAL binary part, for the timer     */

    int done = 0;

    avg_stop_reason = AVG_STOP_RUNAWAY;   /* until proven otherwise        */
    avg_flags = 0;
    avg_time_units = 0.0;
    avg_cycles = 8;                 /* VGGO lead-in: one idle PROM tick    */

    while (!done && fetched < AVG_MAXFETCH) {
        uint16_t w0, w1 = 0;
        int op, dx = 0, dy = 0, z = -1;

        if (!avg_fetch(pc, &w0)) {
            avg_stop_reason = AVG_STOP_BADFETCH;
            break;
        }
        pc++;
        fetched++;

        op = (w0 >> 13) & 7;
        switch (op) {

        case OP_VCTR:                       /* two words: dy then dx|z     */
            if (!avg_fetch(pc, &w1)) {
                avg_stop_reason = AVG_STOP_BADFETCH;
                done = 1;
                break;
            }
            pc++;
            fetched++;
            dy = s13(w0);
            dx = s13(w1);
            z = (w1 >> 13) & 7;
            avg_cycles += 8u * 8u + (uint32_t)(0x8000 -
                avg_timer_reg(avg_norm_shifts((uint16_t)(w0 & 0x1FFF),
                                              (uint16_t)(w1 & 0x1FFF)),
                              bin_scale, 0));
            goto draw;

        case OP_SVEC:                       /* one word, half-resolution   */
            dx = s5(w0) * 2;
            dy = s5((uint16_t)(w0 >> 8)) * 2;
            z = (w0 >> 5) & 7;
            avg_cycles += 6u * 8u + (uint32_t)(0x100 -
                (avg_timer_reg(avg_norm_shifts(
                                   (uint16_t)(((w0 >> 8) & 0x1F) << 8),
                                   (uint16_t)((w0 & 0x1F) << 8)),
                               bin_scale, 1) & 0xFF));
        draw: {
                double fdx = (double)dx * scale;
                double fdy = (double)dy * scale;
                double ax = (fdx < 0.0) ? -fdx : fdx;
                double ay = (fdy < 0.0) ? -fdy : fdy;
                int lum = (z == 1) ? statz : (z << 1);   /* z==1: use STAT */
                double x0 = x, y0 = y;

                x += fdx;
                y += fdy;                   /* +y up; aae negates for raster */
                avg_time_units += (ax > ay) ? ax : ay;   /* aae vector_timer */
                if (seg != NULL)
                    seg((float)x0, (float)y0, (float)x, (float)y, color, lum);
            }
            break;

        case OP_STAT:                       /* $6xxx; bit 12 set = SCAL    */
            avg_cycles += 7u * 8u;
            if (w0 & 0x1000) {              /* SCAL: complemented linear   */
                int bshift = (w0 >> 8) & 7;
                int linear = (~w0) & 0xFF;
                scale = ((double)linear / 256.0) / (double)(1 << bshift);
                bin_scale = bshift;
            } else {                        /* STAT/COLOR ($64xx here)     */
                statz = (w0 >> 4) & 0x0F;
                color = w0 & 0x07;
            }
            break;

        case OP_CNTR:                       /* $8xxx: beam to center       */
            /* w0 & 0xFF is the settling-timer seed: it lands in dvy and
             * the normalizer shifts it up; the centering then runs for
             * 0x8000 - timer master cycles (avg_common_strobe3 OP2 arm). */
            avg_cycles += 5u * 8u + (uint32_t)(0x8000 -
                avg_timer_reg(avg_norm_shifts((uint16_t)(w0 & 0xFF), 0),
                              0, 0));
            x = 0.0;
            y = 0.0;
            break;

        case OP_JSRL: {
            uint16_t a = w0 & 0x1FFF;
            avg_cycles += 5u * 8u;
            if (a == 0) {                   /* reference: target 0 = done  */
                avg_stop_reason = AVG_STOP_JUMP0;
                done = 1;
                break;
            }
            if (sp >= AVG_MAXSTACK - 1) {   /* aae: overflow ends the frame */
                avg_stop_reason = AVG_STOP_STACK_OVER;
                done = 1;
                break;
            }
            stack[sp++] = pc;
            pc = a;
            break;
        }

        case OP_RTSL:
            avg_cycles += 4u * 8u;
            if (sp == 0) {                  /* aae: underflow ends the frame */
                avg_stop_reason = AVG_STOP_STACK_UNDER;
                done = 1;
                break;
            }
            pc = stack[--sp];
            break;

        case OP_JMPL: {
            uint16_t a = w0 & 0x1FFF;
            avg_cycles += 3u * 8u;
            if (a == 0) {
                avg_stop_reason = AVG_STOP_JUMP0;
                done = 1;
                break;
            }
            pc = a;
            break;
        }

        case OP_HALT:
        default:                            /* op is 3 bits: HALT is last  */
            avg_cycles += 2u * 8u;          /* halt visible after 2 ticks  */
            avg_stop_reason = AVG_STOP_HALT;
            done = 1;
            break;
        }
    }
    /* Falling out with !done means the fetch guard tripped: the list never
     * reached a HALT (diagnostic: avg_last_stop() == AVG_STOP_RUNAWAY). */

    if (out_words_walked != NULL)
        *out_words_walked = fetched;
}

void avg_run(void (*seg)(float x0, float y0, float x1, float y1,
                         int color, int lum),
             int *out_words_walked)
{
    avg_run_from(0, seg, out_words_walked);   /* VGGO: word 0 = CPU $2000 */
}

avg_stop avg_last_stop(void)
{
    return avg_stop_reason;
}

int avg_last_flags(void)
{
    return avg_flags;
}

double avg_draw_time_units(void)
{
    return avg_time_units;
}

uint32_t avg_frame_cycles(void)
{
    return avg_cycles;
}

double avg_frame_time_ms(void)
{
    return (double)avg_cycles * (1000.0 / AVG_MASTER_HZ);
}
